edx_event_bus_kafka/internal/producer.py

```diff
@@ -191,16 +194,23 @@ def send(
               string naming the dictionary keys to descend)
             event_data: The event data (kwargs) sent to the signal
         """
-        event_key = extract_event_key(event_data, event_key_field)
-        headers = {EVENT_TYPE_HEADER_KEY: signal.event_type}
-
-        key_serializer, value_serializer = get_serializers(signal, event_key_field)
-        key_bytes = key_serializer(event_key, SerializationContext(topic, MessageField.KEY, headers))
-        value_bytes = value_serializer(event_data, SerializationContext(topic, MessageField.VALUE, headers))
-
-        self.producer.produce(
-            topic, key=key_bytes, value=value_bytes, headers=headers, on_delivery=on_event_deliver,
-        )
+        try:
+            event_key = extract_event_key(event_data, event_key_field)
+            headers = {EVENT_TYPE_HEADER_KEY: signal.event_type}
+
+            key_serializer, value_serializer = get_serializers(signal, event_key_field)
+            key_bytes = key_serializer(event_key, SerializationContext(topic, MessageField.KEY, headers))
+            value_bytes = value_serializer(event_data, SerializationContext(topic, MessageField.VALUE, headers))
+
+            self.producer.produce(
+                topic, key=key_bytes, value=value_bytes, headers=headers, on_delivery=on_event_deliver,
+            )
+        except BadConfigurationException as bce:
+            logger.error(LOG_FORMAT_STRING.format(str(event_data), bce))
+        except MissingKeyException as mke:
+            self._send_to_error_topic(topic=topic, event_data=event_data)
+        except:
+            self._send_to_error_topic(topic=topic, event_data=event_data, event_key=extract_event_key(event_data, event_key_field))
```

Comment:
Questions:
1. Will this be returning https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#kafkaerror? Might we need to investigate details of the error?
2. Do we know what types of exceptions could happen here vs in `on_event_deliver`? Does it matter?
3. Should we send exception information like in https://docs.confluent.io/cloud/current/connectors/dead-letter-queue.html#view-the-dlq error headers?
4. Should we rename to `_send_to_error_topic_or_log`? I was confused at first how we were going to get to Kafka if we just had problems getting to Kafka. I _think_ the plan here is just to try, and if it doesn't work, then we log?

Reply:
For 1/2, so far the only remotely helpful information I can find is https://github.com/confluentinc/confluent-kafka-python/blob/master/tests/test_KafkaError.py , but it's not actually as helpful as I would like it to be. It's clear some things raise and some things just pass along the error but I'm having trouble distinguishing them.
3. yes, probably a good idea
4. i put very little thought into method names, i figured they would be cleaned up in the real thing.

Reply:
Actually these tests are slightly more useful, but only slightly https://github.com/confluentinc/confluent-kafka-python/blob/master/tests/test_Producer.py